Tooltip: add aria-describedby to anchor only if not redundant

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Related to #65888

This PR makes changes to the `Tooltip` component so that the `aria-describedby` attribute gets added more conservatively

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

As discussed in  #65888, the way `Tooltip` adds an `aria-describedby` attribute can disrupt assistive technology users:

- the tooltip is displayed dynamically, which means that the accessible description is only available when the tooltip is visible;
- the tooltip's description can override an existing description;
- the tooltip's description can be unnecessary if it's a copy of the accessible label;
- in general, tooltips should not be a mean of adding an accessible description (although this drawback is not in the scope of this PR)

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

Implementing @t-hamano 's [suggestion](https://github.com/WordPress/gutenberg/issues/65888#issuecomment-2395827508): the `Tooltip` component checks that there isn't a `aria-describedby` attribute already defined, and that the tooltip text is different from the anchor's accessible label.

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a post or page. -->
<!-- 2. Insert a heading block. -->
<!-- 3. etc. -->

- Make sure that a tooltip anchor has an explicit `aria-describedby` attribute. Inspecting the anchor, make sure that the `aria-describedby` attribute doesn't change when the tooltip is shown.
- Make sure that a tooltip anchor has an explicit `aria-label` attribute, and that the tooltip text is the same as the anchor's `aria-label`. Inspecting the anchor, make sure that the `aria-describedby` attribute is not added to the anchor when the tooltip is shown.

I've added a couple of temporary Storybook examples to help with testing in isolation.


In the editor, you can test this by:
- inserting a block
- moving focus to the block toolbar
- inspect the `button` responsible for triggering the "options" dropdown (the last one in the toolbar)
- notice how, with the changes from this PR applied, when the button receives focus and the tooltip shows, the button does _not_ get an `aria-describedby` attribute added for the tooltip

## Screenshots or screencast <!-- if applicable -->

https://github.com/user-attachments/assets/0126844d-6611-41ce-b97a-ec2b3b5dd928

ciampo committed:
Improve comment

tyxla commented on packages/components/src/tooltip/index.tsx:
```diff
@@ -109,7 +109,10 @@ function UnforwardedTooltip(
 	// the tooltip anchor anymore since 0.4.0, so we need to add it manually.
```
We might want to expand the comment to explain under which circumstances we're adding it.

ciampo replied:
[`2935413` (#65989)](https://github.com/WordPress/gutenberg/pull/65989/commits/293541324b9e3c445da7bb1d59ca13c621d90d29)

tyxla reviewed (commented):
Code-wise this looks good to me 👍 I think it's a good compromise, considering we'll seek a better long-term fix.

Leaving the potential approval to @afercia after confirming it resolves the problem.

afercia commented:
Thanks for working at this fix. I don't have bandwidth to test it now but as long as:
- It prevents the double announcement of label and description that have the same value.
- It prevents from dynamically updating existing `aria-describedby` value

then I'd think it's a reasonable short-term solution until the root problem gets solved. Thanks!

ciampo commented:
I'll add a couple of unit tests and temporary storybook examples to validate the thesis further

ciampo committed:
Add unit tests

ciampo committed:
Add temporary storybook examples

ciampo committed:
CHANGELOG

ciampo commented:
Added unit tests, a couple of temporary Storybook examples, and updated PR description.

Let's give this PR a round of review and merge if it improves the current situation for the 6.7 release.

ciampo committed:
Fix lint errors

t-hamano reviewed (approved):
LGTM! 

I tested screen readers primarily on the post editor and site editor, and there were no issues with double reading of labels or inappropriate reading of descriptions.

Here is an example of that test:

### Block Toolbar


https://github.com/user-attachments/assets/cbcd42d0-97f5-489b-a91c-c1298f0458ef

### Header and Sidebar

https://github.com/user-attachments/assets/a157254b-bc1a-4f7c-bdcc-7523e6c5675f

ciampo commented:
Enabled auto-merging + backporting to WordPress 6.7 release

ciampo committed:
Remove temporary storybook examples

DaniGuardiola commented:
Thanks @ciampo! We should look into removing the patch next, which I believe you created a draft PR for somewhere else. I will look at everything once I'm back from sick leave.

ciampo commented:
> We should look into removing the patch next, which I believe you created a draft PR for somewhere else. I will look at everything once I'm back from sick leave.

Yup, see #66021
